### DISCUSSION TITLE
How to create a table component using material-react-table

### DISCUSSION THREAD

**Discussion by emanuella92gomesr** [KevinVandy/material-react-table#560] [Q&A] [answered]
I want to create a table component using _material-react-table_ but I don't know which data type I should use to define its props (such as columns and data). I will define the props in an interface structure. I am using _Typescript_ as programming language.

I used the _MRT_ColumnDef_ as _column_ type and an array of objects as _data_ type and I know that they are incorrect. The columns can be defined using _useMemo_, for example,

```
 useMemo<MRT_ColumnDef<Employee>[]>(
     //define columns data
 );
```
and the data can be represented as an array of objects.

**Comment by KevinVandy** [comment, marked as answer]
You can define a re-usable MRT component that can take any data, but you will probably want to use a TypeScript generic.

```tsx
//extend MaterialReactTableProps so you don't have to type out 200 props yourself
interface Props<TData extends Record<string, any> = {}> extends MaterialReactTableProps<TData> {
  columns: MRT_ColumnDef<TData>[];
  data: TData[];
 //etc.
}

export const YourCustomTableComponent = <TData extends Record<string, any> = {}>({
  columns,
  data,
  ...rest
}: Props<TData>) => {
  //code... maybe custom columns...
  return (
    <MaterialReactTable
      columns={columns}
      data={data}
      //your prop customizations
      {...rest}
    />
  );
}
```

**Reply by emanuella92gomesr** [reply]
MRT and ColumnDef were not found. Should I import some library? 
They weren't found and no fix was suggested.

**Reply by KevinVandy** [reply]
I had typed this out on my phone, fixed some typos

**Reply by emanuella92gomesr** [reply]
The solution works!! 

Thank you very much!

**Reply by ronnyhaase** [reply]
Hi @KevinVandy !

We're currently having the same issue, so we tried your suggestion after struggling with it for hours on our own 😅 . But with the latest TS / MRT versions, we get this error with your example:

> An interface can only extend an object type or intersection of object types with statically known members.

Can you maybe help? 🙏

**Reply by ronnyhaase** [reply]
Figured it out, without using interface:

```tsx
import { MaterialReactTable, MaterialReactTableProps, MRT_ColumnDef } from 'material-react-table';

type Props<TData extends Record<string, any>> = MaterialReactTableProps<TData> & {
    columns: MRT_ColumnDef<TData>[];
    data: TData[];
    // etc.
};

export const YourCustomTableComponent = <TData extends Record<string, any>>({
    columns,
    data,
    ...rest
}: Props<TData>) => {
    return (
        <MaterialReactTable
            columns={columns}
            data={data}
            // your prop customizations
            {...rest}
        />
    );
};
```
